sockeye/inference.py

```diff
@@ -1279,16 +1274,31 @@ def _beam_search(self,
 
             # (3) get beam_size winning hypotheses for each sentence block separately
             # TODO(fhieber): once mx.nd.topk is sped-up no numpy conversion necessary anymore.
-            if self.context == mx.cpu():
-                scores = scores.asnumpy()  # convert to numpy once to minimize cross-device copying
+            scores = scores.asnumpy()  # convert to numpy once to minimize cross-device copying
+
+            # Each row is a batch of flattened beams
+            # (batch_size, beam_size * target_vocab_size)
+            if t == 1:
+                # First time step they're all identical, we take one beam from each sent
+                # (batch_size, vocab_size)
+                folded_scores = scores[::self.beam_size, :]
+            else:
+                # (batch_size, beam_size * vocab_size)
+                folded_scores = np.reshape(scores, (self.batch_size, self.beam_size * len(self.vocab_target)))
 
+            # Get the scores
+            # Indexes into folded_scores: (batch_size, beam_size)
+            flat_idxs = np.argpartition(folded_scores, range(self.beam_size))[:,:self.beam_size]
+            # Score values: (batch_size, beam_size)
+            folded_value_scores = folded_scores[np.arange(folded_scores.shape[0])[:, None], flat_idxs]
+
+            scores_accumulated[:,0] = folded_value_scores.ravel()
+            (best_hyp_indices[:], best_word_indices[:]) = np.unravel_index(flat_idxs.ravel(), scores.shape)
+
+            # TODO hussed: best_hyp_indices are not offset row-wise
+            # use the previous approach, find a cleaner way
             for sent in range(self.batch_size):
```

Comment:
I was actually starting to do the same thing (batched topk) yesterday. One idea I had was to adapt the indices by broadcast adding a mask we can create once in beam search.